src/api/metrics/async_instrument.zig

```diff
@@ -0,0 +1,326 @@
+pub const ObserveMeasures = *const fn (context: ObservedContext, allocator: std.mem.Allocator) MetricObserveError!MeasurementsData;
+
+/// Returns an instance of an asynchronous instrument by Kind.
+/// The type parameter determines the type of the counter: unsigned integers produce an instrument of Kind .ObservableCounter,
+/// signed integers produce an instrument of Kind .ObservableUpDownCounter.
+pub fn ObservableInstrument(K: Kind) type {
+    switch (K) {
+        .ObservableCounter, .ObservableGauge, .ObservableUpDownCounter => {
+            return struct {
+                const Self = @This();
+
+                allocator: std.mem.Allocator,
+                lock: std.Thread.Mutex = .{},
+                /// List of functions that will produce data points when called.
+                /// Functions are called by the Meter when it observes the instrument (e.g. when Metricreader collects metrics).
+                callbacks: ?[]ObserveMeasures = null,
+                context: ObservedContext,
+
+                pub fn init(allocator: std.mem.Allocator, ctx: ?ObservedContext) Self {
+                    return Self{
+                        .allocator = allocator,
+                        .context = ctx orelse .{},
+                    };
+                }
+
+                pub fn deinit(self: *Self) void {
+                    if (self.callbacks) |c| self.allocator.free(c);
+                }
+
+                /// Attaches a callback to the instrument.
+                /// Of separate callbacks produce data points with equal attributes, only the last
+                /// observation is kept, using the order of registration.
+                /// All callbacks are expected to return a MeasurementsData with consistent type.
+                /// If different callbacks return different types, an error is returned when observing them.
+                pub fn registerCallback(self: *Self, callback: ObserveMeasures) !void {
+                    self.lock.lock();
+                    defer self.lock.unlock();
+
+                    if (self.callbacks) |c| {
+                        var new_callbacks = try self.allocator.alloc(ObserveMeasures, c.len + 1);
+                        std.mem.copyForwards(ObserveMeasures, new_callbacks, c);
+                        new_callbacks[c.len] = callback;
+                        self.callbacks = new_callbacks;
+                        self.allocator.free(c);
+                    } else {
+                        self.callbacks = try self.allocator.alloc(ObserveMeasures, 1);
+                        self.callbacks.?[0] = callback;
+                    }
+                }
+
+                fn observe(self: *Self, allocator: std.mem.Allocator) MetricObserveError!?MeasurementsData {
+                    self.lock.lock();
+                    defer self.lock.unlock();
+
+                    if (self.callbacks) |c| {
+                        var m = try allocator.alloc(MeasurementsData, c.len);
```

Comment:
Could this be allocated in stack?

Reply:
That would be good, although in order to be on the stack, the array must have compile-time known size.
Since the number of callbacks is known at runtime, we have to allocate on the heap.

We _could_ make the array of callbacks field `comptime`, but we would have to put un upper bound to it, and do not allow `registerCallback` to be called at runtime (marking it also as `comptime`), but that would violate the OTel spec.